### PR TITLE
Normals for physics/collision shapes

### PR DESCRIPTION
**Pull request by mircokroon**
Current:
image

Shaded:
image

Normals:
image

### CODE REVIEW

File: GUI/Types/Renderer/OctreeDebugRenderer.cs [hunk 1]
```diff
@@ -25,6 +25,33 @@ public static void BindDefaultShaderLayout(int vao, int shaderProgram)
...
+            var positionAttributeLocation = GL.GetAttribLocation(shaderProgram, "aVertexPosition");
+            var normalAttributeLocation = GL.GetAttribLocation(shaderProgram, "aVertexNormal");
+            var colorAttributeLocation = GL.GetAttribLocation(shaderProgram, "aVertexColor");
+
+            GL.EnableVertexArrayAttrib(vao, positionAttributeLocation);
+            GL.EnableVertexArrayAttrib(vao, normalAttributeLocation);
+            GL.EnableVertexArrayAttrib(vao, colorAttributeLocation);
+
+            GL.VertexArrayAttribFormat(vao, positionAttributeLocation, 3, VertexAttribType.Float, false, 0);
+            GL.VertexArrayAttribFormat(vao, normalAttributeLocation, 3, VertexAttribType.Float, false, sizeof(float) * 3);
```
Comment (xPaw) [inline GUI/Types/Renderer/OctreeDebugRenderer.cs:46]: Make normal come after color, so you can call `SimpleVertex.BindDefaultShaderLayout` for setting up position/color.

File: GUI/Types/Renderer/Shaders/basic_shape.frag [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+in vec3 camPos;
+
+out vec4 outputColor;
+
+void main(void) {
+    if (vtxNormal == vec3(0, 0, 0)) {
+        outputColor = vtxColor;
+    } else {
+        vec3 viewDir = normalize(vtxPos - camPos);
+        outputColor = vec4(CalculateFullbrightLighting(vtxColor.rgb, vtxNormal, viewDir) * 2, .8);
```
Comment (xPaw) [inline GUI/Types/Renderer/Shaders/basic_shape.frag:18]:
```suggestion
        outputColor = vec4(CalculateFullbrightLighting(vtxColor.rgb, vtxNormal, viewDir) * 2, vtxColor.a);
```

**Comment by kristiker** [comment]
I prefer the lower opacity.

**Comment by mircokroon** [comment]
Reduced it a bit at, I think lowering it much more makes it harder to read again 

image

File: GUI/Types/Renderer/PhysSceneNode.cs [hunk 3]
```diff
@@ -177,23 +190,34 @@ public static IEnumerable<PhysSceneNode> CreatePhysSceneNodes(Scene scene, PhysA
 
                     // meshes can be large, so ensure capacity up front
                     verts[collisionAttributeIndex].EnsureCapacity(baseVertex + vertices.Length);
-                    inds[collisionAttributeIndex].EnsureCapacity(inds[collisionAttributeIndex].Count + triangles.Length * 6);
+                    inds[collisionAttributeIndex].EnsureCapacity(inds[collisionAttributeIndex].Count + triangles.Length * 6); // TODO account for duplication from normals
```
Comment (xPaw) [inline GUI/Types/Renderer/PhysSceneNode.cs:193]: Can you fix this pls?

Comment (mircokroon) [reply GUI/Types/Renderer/PhysSceneNode.cs]: Actually was already correct, removed the comment

**Review by xPaw — APPROVED** [review]

File: GUI/Types/Renderer/PhysSceneNode.cs [hunk 4]
```diff
@@ -7,31 +9,72 @@ namespace GUI.Types.Renderer
...
+                    indsSizes[sphere.CollisionAttributeIndex] += hemisphereTriangles * 6;
+                }
+                foreach (var capsule in shape.Capsules)
+                {
+                    indsSizes[capsule.CollisionAttributeIndex] += capsuleTriangles * 6;
+                    vertsSizes[capsule.CollisionAttributeIndex] += hemisphereVerts * 2;
+                }
+                foreach (var hull in shape.Hulls)
+                {
+                    var numTriangles = hull.Shape.GetEdges().Length - hull.Shape.GetFaces().Length * 2;
```
Comment (xPaw) [inline GUI/Types/Renderer/PhysSceneNode.cs:59]: I verified that calculated sizes are correct, however calling GetEdges and GetTriangles is less than ideal unfortunately.

**Pull request merged**